### PR TITLE
chore: bump graphql and graphql-compose major versions

### LINKED ISSUE #25906: Support new graphql major (15)
## Description

[graphql 15](https://github.com/graphql/graphql-js/releases/tag/v15.0.0) has been out for a while now (3.5 months), and some libs are starting to upgrade. As a result, today I got this error message:

```
 Error: Cannot use GraphQLDirective "@cacheControl" from another module or realm.
  Ensure that there is only one instance of "graphql" in the node_modules
  directory. If different versions of "graphql" are the dependencies of other
  relied on modules, use "resolutions" to ensure only one version is installed.
  https://yarnpkg.com/en/docs/selective-version-resolutions
  Duplicate "graphql" modules cannot be used at the same time since different
  versions may have different capabilities and behavior. The data from one
  version used in the function from another could produce confusing and
  spurious results.
```

Pinning graphql to 15.3.0 in resolutions makes gatsby crash when running `gatsby develop`.

### Steps to reproduce

Run any gatsby projet with resolutions set to 15.3.0 for graphql.

### Expected result

Gatsby should support the latest graphql major.

### Actual result

Error on build.

### Environment

```
  System:
    OS: macOS 10.15.5
    CPU: (4) x64 Intel(R) Core(TM) i7-7567U CPU @ 3.50GHz
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 14.4.0 - /var/folders/mt/71kkzssd3fq24s3x3y6wpc4h0000gn/T/yarn--1595324137736-0.6757048113533806/node
    Yarn: 1.22.4 - /var/folders/mt/71kkzssd3fq24s3x3y6wpc4h0000gn/T/yarn--1595324137736-0.6757048113533806/yarn
    npm: 6.14.5 - /usr/local/bin/npm
  Languages:
    Python: 2.7.16 - /usr/local/bin/python
  Browsers:
    Chrome: 84.0.4147.89
    Firefox: 72.0.2
    Safari: 13.1.1

```

Thanks in advance!

### PR DESCRIPTION
**Pull request by vladar**
## Description

This PR bumps `graphql-js` major from v14 to v15 and `graphql-compose` from v6 to v7

Closes #25906

### CODE REVIEW

File: packages/gatsby-plugin-image/src/resolver-utils.ts [hunk 1]
```diff
@@ -77,34 +77,34 @@ export function getGatsbyImageFieldConfig<TSource, TContext>(
...
         description: stripIndent`
-            The image formats to generate. Valid values are AUTO (meaning the same format as the source image), JPG, PNG, WEBP and AVIF. 
+            The image formats to generate. Valid values are AUTO (meaning the same format as the source image), JPG, PNG, WEBP and AVIF.
             The default value is [AUTO, WEBP], and you should rarely need to change this. Take care if you specify JPG or PNG when you do
-            not know the formats of the source images, as this could lead to unwanted results such as converting JPEGs to PNGs. Specifying 
+            not know the formats of the source images, as this could lead to unwanted results such as converting JPEGs to PNGs. Specifying
             both PNG and JPG is not supported and will be ignored. AVIF support is currently experimental.
         `,
-        defaultValue: [`auto`, `webp`],
+        defaultValue: [``, `webp`],
```
Comment (pieh) [inline packages/gatsby-plugin-image/src/resolver-utils.ts:94]: is this change intentional?

Comment (vladar) [reply packages/gatsby-plugin-image/src/resolver-utils.ts:94]: Yes, there is no value `auto` in this enum (there is a key `AUTO` but its value is `""`). And `graphql-js` now throws when it sees non-existent enum value (after graphql/graphql-js#1827)

Comment (vladar) [reply packages/gatsby-plugin-image/src/resolver-utils.ts:94]: I think it was working before by a pure accident 😅

Comment (pieh) [reply packages/gatsby-plugin-image/src/resolver-utils.ts:94]: Huh, ok ;) just thinking about readability - would it make sense to do this other way around - set `AUTO` key to `auto` value instead of empty string? :P

Comment (vladar) [reply packages/gatsby-plugin-image/src/resolver-utils.ts:94]: I was thinking about it but this has bigger potential to introduce some bug. See the way this type is defined is pretty weird:

https://github.com/gatsbyjs/gatsby/blob/b93b8bd90fb450518d974f199a83921b97a3715d/packages/gatsby-transformer-sharp/src/types.ts#L17-L22

`NO_CHANGE` and `AUTO` have the same value (empty string). If you only change `AUTO` - then it will change the behavior in one way or another. I didn't want to mess too much with image plugins in this PR so this was a safer choice.

File: packages/gatsby/src/schema/__tests__/__snapshots__/build-schema.js.snap [hunk 2]
```diff
@@ -1,10 +1,7 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
 exports[`Built-in types includes built-in types 1`] = `
-"\\"\\"\\"Provides default value for input field.\\"\\"\\"
-directive @default(value: JSON!) on INPUT_FIELD_DEFINITION
```
Comment (pieh) [inline packages/gatsby/src/schema/__tests__/__snapshots__/build-schema.js.snap:5]: Removal of this directive comes from `graphql-compose` upgrade, right?

Was it ever possible for schema customization to make use of it? I.e. when defining custom field with args? I.e.:
```
args: {
  blah: `String @default("wat")`
}
```
?

Not that we have to keep it if above is possible, just thinking if this is potential breaking change (to document)

File: packages/gatsby/src/schema/__tests__/__snapshots__/print.js.snap [hunk 3]
```diff
@@ -278,6 +278,14 @@ type OneMoreTest implements Node @dontInfer {
...
 union ThisOrThat = AnotherTest | OneMoreTest
 
+type Inline {
+  foo: Nested
+}
+
+input Baz {
+  qux: Boolean
+}
+
```
Comment (pieh) [inline packages/gatsby/src/schema/__tests__/__snapshots__/print.js.snap:288]: This seems to be change (regression) to schema printing? Those are defined as part of https://github.com/gatsbyjs/gatsby/blob/d86aadd692831c5ca28889eff59740c1fc3ab141/packages/gatsby/src/schema/__tests__/print.js#L99-L122

And this particular test should exclude `InlineTest` type that include those inline types:
https://github.com/gatsbyjs/gatsby/blob/d86aadd692831c5ca28889eff59740c1fc3ab141/packages/gatsby/src/schema/__tests__/print.js#L206-L216

And those types are not referenced anywhere else (afaik) - so they are "dead" types (?)

I personally don't know if they should show up or not (if the change of behaviour is desired) - I'll leave this up to you ;)

File: packages/gatsby/src/schema/__tests__/rebuild-sitepage-type.js [hunk 4]
```diff
@@ -157,8 +161,16 @@ describe(`build and update schema for SitePage`, () => {
         keep: String!
         fields: SitePageFields
       }
+      type SitePageFields {
+        temp: String!
+      }
```
Comment (pieh) [inline packages/gatsby/src/schema/__tests__/rebuild-sitepage-type.js:166]: This is needed because of new `deleteFieldsOfDerivedTypes` function? (As in it was kind of broken that test was working as is without haveing `SitePageFields` definition?)

File: packages/gatsby/src/schema/types/sort.ts [hunk 5]
```diff
@@ -149,7 +150,16 @@ export const getSortInput = <TSource = any, TContext = any>({
...
+  // toInputObjectType() will fail to convert fields of union types, e.g.
+  //   union FooBar = Foo | Bar
+  //   type Baz {
+  //     fooBar: FooBar
+  //   }
+  // Unfortunately there is no way to just exclude such fields from the input type in graphql-compose v7+,
+  // so simply replacing them with booleans as the least confusing option :/
+  const inputTypeComposer = toInputObjectType(typeComposer, {
+    fallbackType: schemaComposer.getSTC(`Boolean`),
+  })
```
Comment (pieh) [inline packages/gatsby/src/schema/types/sort.ts:162]: I miss https://github.com/graphql-compose/graphql-compose/blob/640a0b22cee8892ea9e347bd78f68cfbf732ba62/src/utils/toInputObjectType.js#L86-L88 :

meme

File: packages/gatsby/src/schema/__tests__/__snapshots__/build-schema.js.snap (hunk 2 above)
Comment (vladar) [reply packages/gatsby/src/schema/__tests__/__snapshots__/build-schema.js.snap:5]: Yes, it was removed here: https://github.com/graphql-compose/graphql-compose/commit/aa479327f88a0815232f7fbe9783e85c7542978e

I am not sure if we should include notes about it in our migration guide. It was not documented in Gatsby. In fact, we claimed that there is no built-in way to do this: https://www.gatsbyjs.com/docs/reference/graphql-data-layer/schema-customization#setting-default-field-values

File: packages/gatsby/src/schema/__tests__/__snapshots__/build-schema.js.snap [hunk 6]
```diff
@@ -19,13 +16,13 @@ directive @proxy(from: String!, fromNode: Boolean! = false) on FIELD_DEFINITION
 \\"\\"\\"Infer field types from field values.\\"\\"\\"
 directive @infer(
   \\"\\"\\"Don't add default resolvers to defined fields.\\"\\"\\"
-  noDefaultResolvers: Boolean
+  noDefaultResolvers: Boolean @deprecated(reason: \\"noDefaultResolvers is deprecated, annotate individual fields.\\")
```
[the remark this answers is not in the document]
Comment (vladar) [reply packages/gatsby/src/schema/__tests__/__snapshots__/build-schema.js.snap:19]: Yes, this was deprecated a long time ago. But wasn't printed by `graphql-js`. The new version of `graphql-js` now actually prints deprecated messages for arguments. Also, this argument will be actually deleted in https://github.com/gatsbyjs/gatsby/pull/29367

File: packages/gatsby/src/schema/__tests__/__snapshots__/print.js.snap (hunk 3 above)
Comment (vladar) [reply packages/gatsby/src/schema/__tests__/__snapshots__/print.js.snap:288]: The problem is that `graphql-compose` had changed the order of type traversal. It visits nested types too early now. So those types get added to the schema before the original type is filtered out.

To make it work the same as it used to work is expensive - we'll have to traverse all types and fields and search for usage of every type. For me, this doesn't worth the effort at the moment as it shouldn't affect the actual outcome for the end-user.

File: packages/gatsby/src/schema/__tests__/rebuild-sitepage-type.js [hunk 7]
```diff
@@ -99,7 +99,11 @@ describe(`build and update schema for SitePage`, () => {
 
     fields = Object.keys(schema.getType(`SitePage`).getFields())
     expect(fields.length).toBe(12)
-    expect(fields).toEqual(initialFields.concat(`context`))
+
+    // fix ordering of fields (graphql-compose adds them in the weird order in v7)
+    expect(fields).toEqual(
+      initialFields.filter(f => f !== `fields`).concat(`fields`, `context`)
+    )
```
[the remark this answers is not in the document]
Comment (vladar) [reply packages/gatsby/src/schema/__tests__/rebuild-sitepage-type.js:106]: Yes, it's just for the test only. Sorting makes more sense here, changed it. Thanks 👍

File: packages/gatsby/src/schema/__tests__/rebuild-sitepage-type.js (hunk 4 above)
Comment (vladar) [reply packages/gatsby/src/schema/__tests__/rebuild-sitepage-type.js:161]: Yes, this test was not actually testing anything before 😄

File: packages/gatsby/src/schema/__tests__/__snapshots__/print.js.snap (hunk 3 above)
Comment (pieh) [reply packages/gatsby/src/schema/__tests__/__snapshots__/print.js.snap:288]: I think given how expensive that would be it would be fine to have this "quirk" (as long as we are aware of it and it's not something we missed :) )

File: packages/gatsby/src/schema/__tests__/rebuild-sitepage-type.js (hunk 4 above)
Comment (vladar) [reply packages/gatsby/src/schema/__tests__/rebuild-sitepage-type.js:161]: Heh, I just realized that it was working before because inference metadata does not ignore `CREATE_TYPES` (and adds type with `@dontInfer` to the list of ignored types. This way it was kinda working as expected.

But those definitions were ignored in all `schema.js` related calls. And now that we also have `deleteFieldsOfDerivedTypes` we actually need to fully rebuild the type. And unless we also define `SitePageFields` this `build` call will fail because there is no such type in the schema anymore.

File: packages/gatsby/src/schema/types/sort.ts (hunk 5 above)
Comment (vladar) [reply packages/gatsby/src/schema/types/sort.ts:162]: Yes, this is confusing. But the best approach to it would be to fix `graphql-compose` itself (make it allow the old behavior). I'll work on it after higher priority work is done.

**Review by pieh — APPROVED** [review]

**Review by wardpeet — APPROVED** [review]
Let's dooooo this!!!! 👍

File: packages/gatsby-plugin-image/src/resolver-utils.ts (hunk 1 above)
Comment (ascorbic) [reply packages/gatsby-plugin-image/src/resolver-utils.ts:94]: This was for backward-compatibility reasons.
